python/python/tests/test_async_client.py

```diff
@@ -7951,6 +7953,125 @@ async def test_function_delete_with_routing(
+
+        # create a second client to run fcall
+        test_client = await create_client(
+            request, cluster_mode=cluster_mode, protocol=protocol, timeout=15000
+        )
+
+        # call fcall to run the function
+        # make sure that fcall routes to a primary node, and not a replica
+        # if this happens, function_kill and function_stats won't find the function and will fail
+        primaryRoute = SlotKeyRoute(SlotType.PRIMARY, lib_name)
+
+        async def endless_fcall_route_call():
+            # fcall is supposed to be killed, and will return a RequestError
+            with pytest.raises(RequestError) as e:
+                if cluster_mode:
+                    await test_client.fcall_ro_route(
+                        func_name, arguments=[], route=primaryRoute
+                    )
+                else:
+                    await test_client.fcall_ro(func_name, arguments=[])
+            assert "Script killed by user" in str(e)
+
+        async def wait_and_function_kill():
+            # it can take a few seconds for FCALL to register as running
+            await asyncio.sleep(3)
+            timeout = 0
+            while timeout <= 5:
+                # keep trying to kill until we get an "OK"
+                try:
+                    result = await redis_client.function_kill()
+                    #  we expect to get success
+                    assert result == "OK"
+                    break
+                except RequestError:
+                    # a RequestError may occur if the function is not yet running
+                    # sleep and try again
+                    timeout += 0.5
+                    await asyncio.sleep(0.5)
+
+        await asyncio.gather(
+            endless_fcall_route_call(),
+            wait_and_function_kill(),
+        )
+
+        # no functions running so we get notbusy error again
+        with pytest.raises(RequestError) as e:
+            assert await redis_client.function_kill()
+        assert "NotBusy" in str(e)
+
+    @pytest.mark.parametrize("cluster_mode", [False, True])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    async def test_function_kill_write_is_unkillable(
+        self, request, cluster_mode, protocol, redis_client: TGlideClient
+    ):
+        min_version = "7.0.0"
+        if await check_if_server_version_lt(redis_client, min_version):
+            return pytest.mark.skip(reason=f"Redis version required >= {min_version}")
+
+        lib_name = f"mylib1C{get_random_string(5)}"
+        func_name = f"myfunc1c{get_random_string(5)}"
+        code = create_lua_lib_with_long_running_function(lib_name, func_name, 10, False)
+
+        # load the library on all primaries
+        assert await redis_client.function_load(code, replace=True) == lib_name.encode()
+
+        # create a second client to run fcall - and give it a long timeout
+        test_client = await create_client(
```

Comment:
probably, you can create this `test_client` inside `endless_fcall_route_call`

Reply:
I thought that as well, but that means I potentially have to sleep longer in the subthread.